### PR TITLE
[CHNL-28039] 2/3 20 Fence Support

### PR DESCRIPTION
**Pull request by dan-peluso**
# Description
Adds distance calculation and filtering logic to support monitoring only the nearest 20 geofences when accounts have 20+ configured. Uses the Haversine formula for accurate great-circle distance calculations between geographic
 coordinates.

## Due Diligence
- [ ] I have tested this on an emulator and/or a physical device.
- [x] I have added sufficient unit/integration tests of my changes.
- [ ] I have adjusted or added new test cases to team test docs, if applicable.
- [x] I am confident these changes are compatible with all Android versions the SDK currently supports.

## Release/Versioning Considerations
- [x] `Patch` Contains internal changes or backwards-compatible bug fixes.
- [ ] `Minor` Contains changes to the public API.
- [ ] `Major` Contains **breaking** changes.
- [ ] Contains readme or migration guide changes.
- [x] This is planned work for an upcoming release.

## Changelog / Code Overview

**Phase 2 of 3** for supporting accounts with 20+ geofences (CHNL-28039).

### Added:
- `GeofenceDistanceCalculator` object - Calculates distances using Haversine formula
- `GeofenceDistanceCalculator.calculateDistance()` - Returns great-circle distance between two coordinates in meters
- `GeofenceDistanceCalculator.filterToNearest()` - Filters and sorts geofences by proximity to a location
- `KlaviyoGeofence.distanceFrom()` extension function - Convenient distance calculation from a geofence
- 13 comprehensive unit tests covering distance calculations and filtering edge cases

### Changed:
- `KlaviyoLocationManager.startSystemMonitoring()` - Now a suspend function with logic to filter to nearest 20 geofences when 20+ are available
- `KlaviyoLocationManager.startSystemMonitoringCallback()` - New wrapper function to handle async filtering in callback contexts
- Updated all call sites (boot restoration, permission changes, geofence sync) to handle async monitoring

### Notes:
- **Full geofence list is stored** in persistent storage (`dataStore`) as before
- **Only nearest 20 are monitored** by the system's LocationServices for x-platform consistency
- Location-based filtering logic is prepared but not yet active (awaiting Phase 1 merge)
- Currently falls back to monitoring first 20 geofences if 20+ are configured
- Phase 3 will integrate with `getCurrentLocation()` from Phase 1 to enable distance-based filtering

## Test Plan

**Unit Tests:**
```bash
./gradlew :sdk:location:testDebugUnitTest --tests "com.klaviyo.location.GeofenceDistanceCalculatorTest"

Test Coverage:
- Distance calculations between real Northeast US locations (Hartford, Newport, Cambridge, Bushwick)
- Edge cases: zero distance, equator, antipodal points
- Filter behavior: empty lists, fewer than limit, exactly at limit, more than limit
- Sorting verification to ensure closest geofences come first
- Identical location handling

Manual Testing (after Phase 1 integration):
1. Configure 25+ geofences for a test account
2. Start geofence monitoring
3. Verify logs show "Filtering X geofences to nearest 20 based on location"
4. Verify only 20 geofences are registered with LocationServices (check system logs)
5. Verify all geofences remain in persistent storage

Related Issues/Tickets

- https://klaviyo.atlassian.net/browse/CHNL-28039 - Android | Support for accounts configuring 20+ geofences
- Part 1 of 3: Get user location (#371)
- Part 2 of 3: Filter to nearest 20 using Haversine (this PR)
- Part 3: Update API to send location parameters and integrate filtering

### CODE REVIEW

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt [hunk 1]
```diff
@@ -288,7 +304,19 @@ internal class KlaviyoLocationManager : LocationManager {
             }
         }
 
-        geofences.map { geofence ->
+        // Filter to nearest 20 if we have more than 20 geofences
+        // TODO: In Phase 3, integrate with getCurrentLocation() from PR #371 to get precise location
```
Comment (evan-masseau) [inline sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: I thought phase 3 would be just the changes to how the fetch geofences API call works, to include the location in query params.
I think it'd be nice to see the new filtering method in use in this PR for local filtering

File: sdk/location/src/main/java/com/klaviyo/location/GeofenceDistanceCalculator.kt [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        userLatitude: Double,
+        userLongitude: Double,
+        limit: Int = 20
+    ): List<KlaviyoGeofence> {
+        if (geofences.isEmpty()) {
+            return emptyList()
+        }
+
+        // Calculate distance for each geofence and sort by distance
+        return geofences
```
[the remark this answers is not in the document]
Comment (evan-masseau) [reply sdk/location/src/main/java/com/klaviyo/location/GeofenceDistanceCalculator.kt]: I take this back, may as well keep it for the consistency of sorting the list.

**Review by evan-masseau — APPROVED** [review]

File: sdk/location/src/test/java/com/klaviyo/location/KlaviyoLocationManagerTest.kt [hunk 3]
```diff
@@ -863,6 +864,174 @@ internal class KlaviyoLocationManagerTest : BaseTest() {
...
+        mockkStatic("kotlinx.coroutines.tasks.TasksKt")
+
+        // Create a mock Location to return
+        val mockLocation = mockk<android.location.Location>(relaxed = true).apply {
+            every { latitude } returns NYC_LAT
+            every { longitude } returns NYC_LNG
+        }
+
+        // Create mock FusedLocationProviderClient and Task
+        val mockLocationClient = mockk<com.google.android.gms.location.FusedLocationProviderClient>()
```
Comment (evan-masseau) [inline sdk/location/src/test/java/com/klaviyo/location/KlaviyoLocationManagerTest.kt]: fully qualified names
